### PR TITLE
Rewrite measure bac to allow multiclass

### LINKED ISSUE #1834: measures: bac and ber inconsistent?
why does bac not allow multiclass?

### PR DESCRIPTION
**Pull request by AlexEngelhardt**
This PR rewrites the bac measure and the measureBAC function:

- The implementation in bac$fun and measureBAC was redundant. Now bac$fun uses measureBAC
- The balanced accuracy is not anymore computed with the TPR and FPR, but just with the diagonal of the confusion matrix. This makes it able to handle multiclass
- I added two tests: One to make sure this rewrite returns the same result as the old formula, and one to ensure bac works with multiclass now.

Fixes #1834

### CODE REVIEW

**Comment by AlexEngelhardt** [comment]
Travis fails with this error:

    object 'mldr_HL' not found
    1: expect_equal(mldr:::mldr_HL(multi.y, multi.p), measureMultilabelHamloss(multi.y, 
           multi.p)) at testthat/test_base_measures.R:657

This is code I haven't touched, and I can't explain why it broke. Does anyone know what happened there? Is it possible that the new version of `mldr` removed a function `mldr_HL` from it?

**Comment by AlexEngelhardt** [comment]
Merged master and especially #2159 so the tests (will hopefully) pass.

**Comment by AlexEngelhardt** [comment]
Hm, I have no idea why this test failed. Can someone help me here?

      ── 1. Error: FDA properties work (@test_base_fda.R#31)  ────────────────────────
      missing value where TRUE/FALSE needed
      1: expect_silent(train(lrn, subsetTask(tsk, features = 2))) at testthat/test_base_fda.R:31

**Comment by ja-thomas** [comment]
yes, that has nothing to do with your PR, you have to wait until #2173 is resolved and then rebase/merge the master in your branch

**Comment by AlexEngelhardt** [comment]
The tests pass. It seems ready to merge - I added the respective label. (Is that okay, workflow-wise, or is that usually done by a reviewer?)

**Comment by AlexEngelhardt** [comment]
oops - misclicked on "close PR". :)

File: tests/testthat/test_base_measures.R [hunk 1]
```diff
@@ -948,3 +948,25 @@ test_that("setMeasurePars", {
...
   expect_equal(mm$extra.args, list(foo = 1))
 })
+
+test_that("bac works as intended with multiclass tasks", {
+  lrn = makeLearner("classif.rpart")
+  mod = train(lrn, task = iris.task)
+  pred = predict(mod, task = iris.task)
+
+  perf = performance(pred, measures = bac)
+  expect_numeric(perf)
```
Comment (larskotthoff) [inline tests/testthat/test_base_measures.R]: Please check whether the actual performance number is correct here.

**Comment by larskotthoff** [comment]
No, please check by hand one particular example and compare to that hand-computed value.

**Comment by AlexEngelhardt** [comment]
I'm a bit confused :/ With "one particular example", you mean one task, not one observation, right?

If yes, do you mean I should rewrite the following line (line 931 in the latest commit) from the test into a different way of computation that yields the same result?

     expected.bac = mean(diag(table(pred$data$truth, pred$data$response) / table(pred$data$truth, pred$data$truth)))

**Comment by larskotthoff** [comment]
See for example https://github.com/mlr-org/mlr/blob/master/tests/testthat/test_base_measures.R#L214 -- the errors there are hand-computed and compared to.

**Comment by AlexEngelhardt** [comment]
Ah, I see. I changed the test case to the small manual example you linked to. Is this what you had in mind?

Would it make sense to incorporate that test case in the existing large `test_that("check measure calculations"` test case, or should I leave it swapped out there?

**Comment by larskotthoff** [comment]
Thanks, that's fine. Merging.

**Comment by AlexEngelhardt** [comment]
Oh okay, I see. 

I thought that hard-coded numbers are more error-prone than formulas, that's why I just discarded that possibility. But ok - now I know for next time.

**Comment by berndbischl** [comment]
> I thought that hard-coded numbers are more error-prone than formulas, that's why I just discarded that possibility. But ok - now I know for next time.

it really is less error prone as it makes you think about the formula you implemented.

**Pull request merged**